libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1334,28 +1336,112 @@ string YulUtilFunctions::allocationFunction()
+		.render();
+	});
+}
+
+string YulUtilFunctions::zeroComplexMemoryArrayFunction(ArrayType const& _type)
+{
+	solAssert(!_type.baseType()->hasSimpleZeroValueInMemory(), "");
+
+	string functionName = "zero_complex_memory_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(memPtr, allocSize) {
+				let firstSlot := memPtr
+				let dataSize := allocSize
 				<?dynamic>
-				mstore(memPtr, length)
+				firstSlot := add(firstSlot, 32)
+				dataSize := sub(dataSize, 32)
 				</dynamic>
+				for {let i := 0} lt(i, dataSize) { i := add(i, 32) } {
+					mstore(add(firstSlot, i), <zeroValue>())
+				}
 			}
 		)")
 		("functionName", functionName)
-		("alloc", allocationFunction())
-		("allocSize", arrayAllocationSizeFunction(_type))
 		("dynamic", _type.isDynamicallySized())
+		("zeroValue", zeroValueFunction(*_type.baseType()))
 		.render();
 	});
 }
 
+string YulUtilFunctions::allocateMemoryArrayFunction(ArrayType const& _type)
+{
+	solUnimplementedAssert(!_type.isByteArray(), "");
+
+	string functionName = "allocate_and_zero_memory_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		auto const* base = _type.baseType();
+		solUnimplementedAssert(base->sizeOnStack() == 1, "Stacksize not yet implemented!");
+		return Whiskers(R"(
+				function <functionName>(length) -> memPtr {
+					let allocSize := <allocSize>(length)
+					memPtr := <alloc>(allocSize)
+					<zeroArrayFunction>(memPtr, allocSize)
+					<?dynamic>
+					mstore(memPtr, length)
+					</dynamic>
+				}
+			)")
+			("functionName", functionName)
+			("alloc", allocationFunction())
+			("allocSize", arrayAllocationSizeFunction(_type))
+			("zeroArrayFunction", base->hasSimpleZeroValueInMemory() ? zeroSimpleMemoryArrayFunction(_type) : zeroComplexMemoryArrayFunction(_type))
+			("dynamic", _type.isDynamicallySized())
+			.render();
+	});
+}
+
+string YulUtilFunctions::allocateMemoryStructFunction(StructType const& _type)
+{
+	string functionName = "allocate_and_zero_memory_struct_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+		function <functionName>() -> memPtr {
+			memPtr := <alloc>(<allocSize>)
+			let offset := memPtr
+			<#member>
+				mstore(offset, <zeroValue>())
+				offset := add(offset, 32)
+			</member>
+		}
+		)");
+		templ("functionName", functionName);
+		templ("alloc", allocationFunction());
+
+		TypePointers const& members = _type.memoryMemberTypes();
+		templ("allocSize", to_string(members.size() * 32));
+
+		vector<map<string, string>> memberParams(members.size());
+		for (size_t i = 0; i < members.size(); ++i)
+		{
+			solAssert(members[i]->memoryHeadSize() == 32, "");
+			if (auto const* refType = dynamic_cast<ReferenceType const*>(members[i]))
+			{
+				unique_ptr<ReferenceType> refTypeWithLocation = refType->copyForLocation(DataLocation::Memory, true);
```

Comment:
ReferenceType::copyForLocationIfReference

Reply:
Actually I'm wondering if we should not change `StructType::memoryMemberTypes` instead.

Reply:
Yea that could be a good idea